core/src/session.rs

```diff
@@ -614,50 +505,271 @@ impl Drop for SessionInternal {
-        let session = match self.1.try_upgrade() {
+        use KeepAliveState::*;
+
+        let session = match self.session.try_upgrade() {
             Some(session) => session,
             None => return Poll::Ready(Ok(())),
         };
 
+        // Process all messages that are immediately ready
         loop {
-            let (cmd, data) = match ready!(self.0.try_poll_next_unpin(cx)) {
-                Some(Ok(t)) => t,
-                None => {
+            match self.as_mut().project().stream.try_poll_next(cx) {
+                Poll::Ready(Some(Ok((cmd, data)))) => {
+                    let result = self.as_mut().dispatch(&session, cmd, data);
+                    if let Err(e) = result {
+                        debug!("could not dispatch command: {}", e);
+                    }
+                }
+                Poll::Ready(None) => {
                     warn!("Connection to server closed.");
                     session.shutdown();
                     return Poll::Ready(Ok(()));
                 }
-                Some(Err(e)) => {
+                Poll::Ready(Some(Err(e))) => {
                     error!("Connection to server closed.");
                     session.shutdown();
                     return Poll::Ready(Err(e));
                 }
-            };
+                Poll::Pending => break,
+            }
+        }
 
-            if let Err(e) = session.dispatch(cmd, data) {
-                debug!("could not dispatch command: {}", e);
+        // Handle the keee-alive sequence, returning an error when we haven't received a
+        // Ping/PongAck for too long.
+        //
+        // The expected keepalive sequence is
+        // - Server: Ping
+        // - wait 60s
+        // - Client: Pong
+        // - Server: PongAck
+        // - wait 60s
+        // - repeat
+        //
+        // This means that we silently lost connection to Spotify servers if
+        // - we don't receive a Ping 60s after the last PongAck, or
+        // - we don't receive a PongAck immediately after our Pong.
+        //
+        // Currently, we add a safety margin of 5s to these expected deadlines.
+        let mut this = self.as_mut().project();
+        if let Poll::Ready(()) = this.timeout.as_mut().poll(cx) {
+            match this.keep_alive_state {
+                ExpectingPing | ExpectingPongAck => {
+                    if !session.is_invalid() {
+                        session.shutdown();
+                    }
+                    // TODO: Optionally reconnect (with cached/last credentials?)
+                    return Poll::Ready(Err(io::Error::new(
+                        io::ErrorKind::TimedOut,
+                        "session lost connection to server",
+                    )));
+                }
+                PendingPong => {
+                    trace!("Sending Pong");
+                    let _ = session.send_packet(PacketType::Pong, vec![0, 0, 0, 0]);
```

Comment:
It seems to be possible to flush the stream in principle (i.e. await the `<Framed<TcpStream> as SinkExt<_>>::poll` method). However, as the `sender_task` is implemented right now, there's only a one-directional channel (`tx_connection`) to send data. It would require a redesign of the `Session` to handle flushing.

I'd argue that it's a bit too much to put in this PR. In addition, flushing behavior (or rather the lack thereof) is the same as it was before this PR, so it's rather an orthogonal issue to the changed keep-alive sequence.

In general, I know far from enough on TCP and its implementations to tell what is required to make flushing work: For example, there's the `TCP_NODELAY` socket option (i.e. `socket.set_nodelay(true)`. Would we need to set that additionally?

The 5 second delay in particular is pretty arbitrary, maybe it would make sense to raise that a bit to account for short network drop-outs? If I remember correctly, I've observed the `PongAck` within a few hundred milliseconds after sending the `Pong` on a wired connection.

Reply:
I agree with everything you've said there. And I've also seen some relatively large pong-ack delays (which could be due to us or them or just network, I never investigated). Arguably unnecessary to invalidate the session over missing pong-ack. The next missing ping will sort things out for us anyway.

Reply:
Ok, so I guess the remaining question is under which condition exactly we should time out:

1. no Ping (120s + x) after the last Ping
2. no Ping (60s + x) after sending the last Pong
3. no Ping (60s + x) after receiving the last PongAck
4. no PongAck x after sending the last Pong

where x accounts for any extra latency in the network round-trip and processing of keep-alive events.

This PR implements 3 + 4, whereas the old behaviour is 1. To be honest, I have no idea what would be best (presumably, that would depend on how the server handles timing of these events). Is your suggestion 1, 2 or 3 here?

In fact, I've also seen a few connection resets (spaced several hours apart) with this PR applied, but I didn't investigate exactly where they originated. Maybe, 5s is in fact too little margin for our timeouts?

Reply:
Two questions:
1. Does 3 + 4 not also imply 2?
2. What do you mean with "old behaviour" - where, in #1357 or in current `dev`?

First thought is that I agree with @kingosticks to not really care about the PongAcks.

Reply:
> Two questions:
> 
> 1. Does 3 + 4 not also imply 2?

If there's not much latency, essentially yes. I think it's not quite the same regarding how the tolerance `x` adds up and from which points exactly the timeouts are counted if some event (say the PongAck) arrives with significant delay (but I didn't really think this through in detail).

> 2. What do you mean with "old behaviour" - where, in [Rework session keep-alive logic #1357](https://github.com/librespot-org/librespot/pull/1357) or in current `dev`?

Current dev. #1357 has the same behavior as this PR; only the implementation is different.

> First thought is that I agree with @kingosticks to not really care about the PongAcks.

My first thought was to use as much information about the state of the connection as possible, which is how I ended up here (i.e. time out shortly after missing PongAck instead of waiting for a next Ping). You may be right that this overcomplicates things.

In the end, this might mean that just taking the first commit of https://github.com/librespot-org/librespot/pull/1357 and dropping all other commits of #1357 and #1359 is the way to go for simplicity.

Reply:
librespot-java also ignores PongAck, but if I understand the code correctly, it behaves like current librespot(-rs) `dev` and send the `Pong` immediately after receiving a `Ping`.

Reply:
I really don't get why it's an issue only for some accounts. I disabled my pongs entirely as an experiment and while i did then manage to get connection reset, it was only a couple over a long period.

Reply:
> In the end, this might mean that just taking the first commit of https://github.com/librespot-org/librespot/pull/1357 and dropping all other commits of #1357 and #1359 is the way to go for simplicity.

Though I stand by my earlier statement that correctness trumps simplicity, if the end result is the same then I would also say that Occam's razor applies. 

> I really don't get why it's an issue only for some accounts. I disabled my pongs entirely as an experiment and while i did then manage to get connection reset, it was only a couple over a long period.

Spotify is the weirdest; some APs respond and behave differently. I remember for some users there's also no autoplay flag being reported in the XML of the initial connection, yet for 99% it is. I would not even know how to get the autoplay state without it. 

Due to the GeoIP thing it's hard if at all possible to make a manual connection to such APs unless you are actually there.